core/src/data_sources/splitter.rs

```diff
@@ -79,37 +81,111 @@ impl Splitter for BaseV0Splitter {
         let re = Regex::new(r"\s+").unwrap();
         let clean = re.replace_all(text, " ");
         let clean = clean.trim();
-
         // Get the embedder and initialize.
         let mut embedder = provider(provider_id).embedder(model_id.to_string());
         embedder.initialize(credentials).await?;
 
         // Encode the clean text.
         let encoded = embedder.encode(clean).await?;
+        let mut decoded = vec![];
+
+        let mut start = 0;
+        while start < encoded.len() {
+            let end = cmp::min(start + max_chunk_size - 1, encoded.len() - 1);
 
-        // Split the encoded text into chunks of size max_chunk_size.
-        let mut chunks = Vec::new();
-        let mut chunk = Vec::new();
-        let mut chunk_size = 0;
-        for token in encoded {
-            chunk.push(token);
-            chunk_size += 1;
-            if chunk_size >= max_chunk_size {
-                chunks.push(chunk);
-                chunk = Vec::new();
-                chunk_size = 0;
+            match decode_chunk_with_remainder(&embedder, &encoded[start..=end]).await {
+                Ok((chunk_decoded, chunk_remainder)) => {
+                    
+                    if chunk_remainder.is_some() {
+                        let chunk_remainder_value = chunk_remainder.unwrap();
+                        if chunk_remainder_value.len() == max_chunk_size {
+                            return Err(anyhow!("The remainder chunk is the same size as the max chunk size - can't keep decoding. max_chunk_size: {}", max_chunk_size));
+                        }
+                        start += (end - start) + 1 - chunk_remainder_value.len();
+                    } else {
+                        start += (end - start) + 1;
+                    }
+                    
+                    if chunk_decoded.is_some() {
+                        decoded.push(chunk_decoded.unwrap());
+                    }
+                },
+                Err(e) => {
+                    return Err(e);
+                }
             }
         }
-        if chunk.len() > 0 {
-            chunks.push(chunk);
-        }
 
-        // Decode the chunks in parallel.
-        let mut futures = Vec::new();
-        for chunk in chunks {
-            futures.push(embedder.decode(chunk));
+        async fn decode_chunk_with_remainder(
+            embedder: &Box<dyn Embedder + Sync + Send>,
+            chunk: &[usize],
+        ) -> Result<(Option<String>, Option<Vec<usize>>), anyhow::Error> {
```

Comment:
you don't need `anyhow::Error` here. You can just do `Result<(Option<String>, Option<Vec<usize>>)>`